### PR TITLE
ARROW-5828: [C++] Add required Protocol Buffers versions check

### PR DESCRIPTION
**Pull request by kou**
This works Protobuf_SOURCE=AUTO well on environments that have old
Protocol Buffers.

### CODE REVIEW

File: cpp/cmake_modules/ThirdpartyToolchain.cmake [hunk 1]
```diff
@@ -180,6 +180,19 @@ macro(resolve_dependency DEPENDENCY_NAME)
...
 
+macro(resolve_dependency_with_version DEPENDENCY_NAME REQUIRED_VERSION)
+  if(${DEPENDENCY_NAME}_SOURCE STREQUAL "AUTO")
+    find_package(${DEPENDENCY_NAME} ${REQUIRED_VERSION} MODULE)
+    if(NOT ${${DEPENDENCY_NAME}_FOUND})
+      build_dependency(${DEPENDENCY_NAME})
+    endif()
+  elseif(${DEPENDENCY_NAME}_SOURCE STREQUAL "BUNDLED")
+    build_dependency(${DEPENDENCY_NAME})
+  elseif(${DEPENDENCY_NAME}_SOURCE STREQUAL "SYSTEM")
```
Comment (kszucs) [inline cpp/cmake_modules/ThirdpartyToolchain.cmake]: Missing `CONDA` source.

Comment (kou) [reply cpp/cmake_modules/ThirdpartyToolchain.cmake]: `CONDA` source is resolved at the first: https://github.com/apache/arrow/blob/master/cpp/cmake_modules/ThirdpartyToolchain.cmake#L89-L99
Then we can process as `SYSTEM` source.

Comment (kszucs) [reply cpp/cmake_modules/ThirdpartyToolchain.cmake]: Indeed.

**Review by kszucs — APPROVED** [review]
LGTM

**Pull request closed without merging**